Clear Database Test Rule

Sloy opened the pull request:
This Rule works similarly to #24, but for SQLite databases.

It fetches all databases in the app, obtains all tables from them, and performs a "delete from" [without where](https://www.youtube.com/watch?v=i_cVJgIz_Cs).

I decided to do this instead of a DROP because I think it might break the app, not sure though.

This should be tested against different scenarios and sdk versions. I've seen in [Stetho](https://github.com/facebook/stetho/blob/master/stetho/src/main/java/com/facebook/stetho/inspector/database/DefaultDatabaseConnectionProvider.java) that they have different ways of opening databases depending on the sdk version. Wonder why.

TODO:
- [x] Check on different sdk versions, and fix any issues
- [x] Document on the javadoc and readme
- [x] (Won't do) _Maybe_ join with ClearPreferencesRule into one single rule for clearing app data.

Sloy commented:
One Issue found: It doesn't work if the database hasn't been created prior to the test (which will be the usual case). We must only try to delete databases if they exist. 🙌 

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.schibsted.spain.barista.sample/com.schibsted.spain.barista.sample.DatabaseActivity}: android.database.sqlite.SQLiteException: no such table: User (code 1): , while compiling: SELECT name FROM User
#################################################################
Error Code : 1 (SQLITE_ERROR)
Caused By : SQL(query) error or missing database.
(no such table: User (code 1): , while compiling: SELECT name FROM User)
```

rocboronat commented:
Great work! Let's see how it progress!

By the way @Sloy , what do you think about having one Rule for every kind of thing to be remoded, and another one that just joins all of them? I think that having different rules will make them simpler. So, if we want to ease using all of them at the same tame (that will be the most common way to work with Barista), we can just create a Rule to join them. And... well, we can talk about it later, let's focus on this particular Rule :·D

Sloy commented:
That was my original idea. But I wasn't too sure. I thought about having just one for everything for simplicity, but if you think having both can be useful, let's go with it. I don't have a strong opinion on neither.

rocboronat commented:
@Sloy yes... the Rule that clears the Database needs 80 lines. The SharedPrefs one, 80 lines more I guess. Less have them smaller as possible, and create a new Rule to join these Rules.

By the way, at some point of the future, we will need to remove all .jpg and .png pictures of the filesystem (or whatever other extension) so... let's keep them isolated.

Sloy commented:
One rule to rule them all!! 💍

rocboronat commented:
:joy:

Sloy commented:
🍏  The issue I commented about clearing databases that don't exist is now fixed.
I changed the way of finding databases and opening them. I had to take into account the shadow database files like those ended in "-journal".

rocboronat commented on library/src/main/java/com/schibsted/spain/barista/cleardata/ClearDatabaseRule.java:
```diff
@@ -0,0 +1,112 @@
+ */
+public class ClearDatabaseRule implements TestRule {
+
+  private static final String[] UNINTERESTING_FILENAME_SUFFIXES = new String[] {
```
Niiice!

rocboronat commented on sample/src/main/java/com/schibsted/spain/barista/sample/DatabaseActivity.java:
```diff
@@ -0,0 +1,71 @@
+    writableDatabase.close();
+  }
+
+  private static class OpenHelper extends SQLiteOpenHelper {
```
We could call it `OpenDatabaseHelper`. When I seen the dependencies of this Activity, I went crazy: What's this class?

Several horses and monkey died during my madness.

rocboronat commented on sample/src/main/java/com/schibsted/spain/barista/sample/DatabaseActivity.java:
```diff
@@ -0,0 +1,71 @@
+  private static class OpenHelper extends SQLiteOpenHelper {
+
+    public OpenHelper(Context context) {
+      super(context, "mydatabase", null, 1);
```
What means that `1`?

rocboronat commented on library/src/main/java/com/schibsted/spain/barista/cleardata/ClearDatabaseRule.java:
```diff
@@ -0,0 +1,112 @@
+    return tidiedList;
+  }
+
+  private static String removeSuffix(String str, String[] suffixesToRemove) {
```
What do you think about calling this method `filterUninterestingSuffixes`? And I suggest to avoid passing the list of suffixest to remove. It seems that they will be always the same:  `UNINTERESTING_FILENAME_SUFFIXES`.

rocboronat commented on library/src/main/java/com/schibsted/spain/barista/cleardata/ClearDatabaseRule.java:
```diff
@@ -0,0 +1,112 @@
+
+  private void clearData() {
+    for (File dbFile : tidyDatabaseList(getDatabaseFiles())) {
+      SQLiteDatabase database = performOpen(dbFile, 0);
```
What means this `0`?

rocboronat commented on library/src/main/java/com/schibsted/spain/barista/cleardata/ClearDatabaseRule.java:
```diff
@@ -0,0 +1,112 @@
+   * @param databaseFiles Raw list of database files.
+   * @return Tidied list with shadow databases removed.
+   */
+  private static List<File> tidyDatabaseList(List<File> databaseFiles) {
```
This method name sounds strange... Don't know. I don't have a suggestion, by the way :·)

rocboronat commented:
There are some comments but... great work!

Sloy commented:
Rebased an squashed, much of the code on the rule has been changed @rocboronat.
I will check out your comments on the rest of the code later :)

rocboronat commented:
Nice, @Sloy ! Check my comments and go on applying the suggestion or merging :·)

Sloy commented:
Travis likes it. I tried a local build with InfoJobs project and everything works as expected.
Any final words @rocboronat?

rocboronat commented:
Ave, @Sloy , morituri te salutant! Great work there :·) It's merge time! Feel free to release a version if you need it! :·)
